plugins: each plugin instance now has its own state

Insei opened the pull request:
Now each plugin instance now has its own state

If we use coreDHCP as a library and run multiple servers, we can't separate configs variables for each server. This PR fix this problem.

Signed-off-by: Dmitrii Aleksandrov <goodmobiledevices@gmail.com>

Natolumin commented on plugins/netmask/plugin.go:
```diff
@@ -40,17 +40,19 @@ func setup4(args ...string) (handler.Handler4, error) {
 //Handler4 handles DHCPv4 packets for the netmask plugin
-func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
-	resp.Options.Update(dhcpv4.OptSubnetMask(netmask))
+func (p pluginState) Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
```
Also a value receiver

Natolumin commented on plugins/netmask/plugin_test.go:
```diff
@@ -25,7 +25,7 @@ func TestCheckValidNetmask(t *testing.T) {
 func TestHandler4(t *testing.T) {
 	// set plugin netmask
-	netmask = net.IPv4Mask(255, 255, 255, 0)
+	pState := &pluginState{netmask: net.IPv4Mask(255, 255, 255, 0)}
```
Can we make this with the setup function ?

Natolumin commented on plugins/netmask/plugin_test.go:
```diff
@@ -35,17 +35,26 @@ func TestHandler4(t *testing.T) {
-	assert.EqualValues(t, netmask, net.IPv4Mask(255, 255, 255, 0))
+	handler, err := setup4("255.255.255.0")
+
+	// prepare DHCPv4 request
+	req := &dhcpv4.DHCPv4{}
+	resp := &dhcpv4.DHCPv4{
+		Options: dhcpv4.Options{},
+	}
+
+	result, stop := handler(req, resp)
+	assert.Same(t, result, resp)
+	assert.False(t, stop)
+	assert.EqualValues(t, resp.Options.Get(dhcpv4.OptionSubnetMask), net.IPv4Mask(255, 255, 255, 0))
```
This completely changes the purpose of the test - this test should check that setup4 creates the internal state we expect, it shouldn't call the handler function

Insei replied:
Ok

Natolumin commented on plugins/prefix/plugin.go:
```diff
@@ -77,8 +77,8 @@ type lease struct {
 
-// Handler holds state of allocations for the plugin
-type Handler struct {
+// PluginState holds state of allocations for the plugin
```
If we're changing this to make it homogeneous across plugins, let's also make it private at the same time

Insei replied:
Allready done

Natolumin commented on plugins/router/plugin.go:
```diff
@@ -6,12 +6,11 @@ package router
 	"github.com/coredhcp/coredhcp/logger"
 	"github.com/coredhcp/coredhcp/plugins"
 	"github.com/insomniacslk/dhcp/dhcpv4"
+	"net"
```
Another line where your code formatter did something weird

Insei replied:
Yep, fixed

Natolumin commented on plugins/router/plugin.go:
```diff
@@ -22,28 +21,29 @@ var Plugin = plugins.Plugin{
 //Handler4 handles DHCPv4 packets for the router plugin
-func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
-	resp.Options.Update(dhcpv4.OptRouter(routers...))
+func (p pluginState) Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
```
another value receiver

Natolumin commented on plugins/staticroute/plugin.go:
```diff
@@ -37,35 +39,35 @@ func setup4(args ...string) (handler.Handler4, error) {
 // Handler4 handles DHCPv4 packets for the static routes plugin
-func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
-	if len(routes) > 0 {
+func (p pluginState) Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
```
another value receiver

Insei replied:
Ok

Natolumin commented on plugins/staticroute/plugin_test.go:
```diff
@@ -38,17 +36,42 @@ func TestSetup4(t *testing.T) {
+
 	// valid route
-	_, err = setup4("10.0.0.0/8,192.168.1.1")
+	handler, err := setup4("10.0.0.0/8,192.168.1.1")
+	result, stop := handler(req, resp)
+	assert.Same(t, result, resp)
+	assert.False(t, stop)
+	table := result.Options.Get(dhcpv4.OptionStaticRoutingTable)
+	routes := dhcpv4.Routes{}
+	if err := routes.FromBytes(table); err != nil {
+		t.Errorf("FromBytes(%v) Unexpected error state: %v", table, err)
+	}
```
This should be a different test, eg `TestHandler4`

Insei replied:
No problem

Natolumin reviewed (commented):
Thanks for the PR! 
That's definitely a direction we want to go into so this is very appreciated. Do you have ideas maybe for tests we could have to avoid regressions on this point ?

I've left a couple notes inline for minor code notes, but this is probably fine to merge with minor changes once I get the time to test it a little

Insei commented on plugins/file/plugin_test.go:
```diff
@@ -17,6 +18,11 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+var pTestState = &pluginState{
+	recLock:       sync.RWMutex{},
+	staticRecords: map[string]net.IP{},
+}
+
```
Yep, we can do this in all tests, but i simple adopt without changes all tests for current model of plugin state.
I do this in next commits.

Insei commented on plugins/file/plugin.go:
```diff
@@ -200,16 +194,24 @@ func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
 
 func setup6(args ...string) (handler.Handler6, error) {
-	h6, _, err := setupFile(true, args...)
+	pState := &pluginState{
+		recLock:       sync.RWMutex{},
+		staticRecords: map[string]net.IP{},
+	}
+	h6, _, err := pState.setupFile(true, args...)
 	return h6, err
 }
 
 func setup4(args ...string) (handler.Handler4, error) {
-	_, h4, err := setupFile(false, args...)
+	pState := &pluginState{
+		recLock:       sync.RWMutex{},
+		staticRecords: map[string]net.IP{},
+	}
+	_, h4, err := pState.setupFile(false, args...)
 	return h4, err
 }
```
Then, I will remove pluginState from all tests and will use only the setup and handler functions.

Insei commented on plugins/mtu/plugin.go:
```diff
@@ -25,26 +24,27 @@ var Plugin = plugins.Plugin{
 
 // Handler4 handles DHCPv4 packets for the mtu plugin
-func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
+func (p pluginState) Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
```
Your project, your rules)

Insei commented on plugins/mtu/plugin.go:
```diff
@@ -7,9 +7,8 @@ package mtu
-	"strconv"
-
 	"github.com/insomniacslk/dhcp/dhcpv4"
+	"strconv"
```
Ok

Insei commented on plugins/mtu/plugin_test.go:
```diff
@@ -21,9 +21,9 @@ func TestAddServer4(t *testing.T) {
 	}
 
-	mtu = 1500
+	pState := &pluginState{mtu: 1500}
```
Ok

Insei commented on plugins/searchdomains/plugin.go:
```diff
@@ -51,27 +52,29 @@ func copySlice(original []string) []string {
 }
 
-func domainSearchListHandler6(req, resp dhcpv6.DHCPv6) (dhcpv6.DHCPv6, bool) {
+func (p pluginState) Handler6(req, resp dhcpv6.DHCPv6) (dhcpv6.DHCPv6, bool) {
```
Ok

Insei commented on plugins/serverid/plugin.go:
```diff
@@ -26,15 +26,18 @@ var Plugin = plugins.Plugin{
-// v6ServerID is the DUID of the v6 server
-var (
+type pluginStateV6 struct {
+	// v6ServerID is the DUID of the v6 server
 	v6ServerID *dhcpv6.Duid
+}
+
+type pluginStateV4 struct {
 	v4ServerID net.IP
-)
+}
```
Yes, you are right

Insei commented on plugins/staticroute/plugin.go:
```diff
@@ -23,11 +23,13 @@ var Plugin = plugins.Plugin{
 func setup4(args ...string) (handler.Handler4, error) {
 	log.Printf("loaded plugin for DHCPv4.")
-	routes = make(dhcpv4.Routes, 0)
+	pState := &pluginState{routes: []*dhcpv4.Route{}}
```
Fixed

Insei commented on plugins/nbp/nbp.go:
```diff
@@ -91,16 +99,17 @@ func setup4(args ...string) (handler.Handler4, error) {
-	opt67 = &obfn
+	pState := &pluginStateV4{
+		opt67: &obfn,
+		opt66: &otsn,
```
Oh eah, sorry, i forgot that dhcpv4.Option is a struct, because dhcpv6.Option is an interface.

Insei commented:
Okay, i think it's ready.
I fixed all the remarks, and fixed some tests.

I also changed the plugin's server_id test, it's not entirely clear why UUID DUID's are tested if they are not supported by the plugin. I can't even call setup method with UUID DUID. Therefore, I redid all the tests for LL DUID and LLT DUID.

Insei commented:
> Thanks for the PR!
> That's definitely a direction we want to go into so this is very appreciated. Do you have ideas maybe for tests we could have to avoid regressions on this point ?

Sorry, i simple use this project. I also don't have any particular performance requirements. I just liked this project more than others, and I took up its revision to fit my needs.

Insei commented:
@Natolumin any update? Can we merge this or no?

dulitz commented:
I have a few more plugins I'd like to upstream. Personally I think this PR goes in the right direction so I'd like to see it merged. But when it's merged I'll have to make changes to my plugins so I'd like to know one way or another.
